File: cli/python/jadevectordb/formatters.py
"""
Output formatters for JadeVectorDB CLI

This module provides formatting functions for different output formats
(JSON, YAML, table, CSV) to improve CLI usability and integration.
"""

import csv
import io
import json
import sys
from typing import Any, Dict, List, Optional

def format_json(data: Any, indent: int = 2) -> str:
    """
    Format data as JSON

    :param data: Data to format
    :param indent: Number of spaces for indentation
    :return: JSON formatted string
    """
    return json.dumps(data, indent=indent)

def format_yaml(data: Any) -> str:
    """
    Format data as YAML

    :param data: Data to format
    :return: YAML formatted string
    """
    try:
        import yaml
        return yaml.dump(data, default_flow_style=False, sort_keys=False)
    except ImportError:
        print("Warning: PyYAML not installed. Install with: pip install PyYAML", file=sys.stderr)
        print("Falling back to JSON output:", file=sys.stderr)
        return format_json(data)

def format_table(data: Any, headers: Optional[List[str]] = None) -> str:
    """
    Format data as a table

    :param data: Data to format (dict, list of dicts, or list of lists)
    :param headers: Optional custom headers for table columns
    :return: Table formatted string
    """
    try:
        from tabulate import tabulate

        # Handle different data types
        if isinstance(data, dict):
            # Single dictionary - show as key-value pairs
            table_data = [[k, v] for k, v in data.items()]
            return tabulate(table_data, headers=headers or ["Key", "Value"], tablefmt="grid")

        elif isinstance(data, list):
            if not data:
                return "No data to display"

            # List of dictionaries - show as table
            if isinstance(data[0], dict):
                if headers is None:
                    headers = list(data[0].keys())
                table_data = [[item.get(h, '') for h in headers] for item in data]
                return tabulate(table_data, headers=headers, tablefmt="grid")

            # List of lists - show as table
            else:
                return tabulate(data, headers=headers or [], tablefmt="grid")

        else:
            # For other types, convert to string
            return str(data)

    except ImportError:
        print("Warning: tabulate not installed. Install with: pip install tabulate", file=sys.stderr)
        print("Falling back to JSON output:", file=sys.stderr)
        return format_json(data)

def format_csv(data: Any, headers: Optional[List[str]] = None) -> str:
    """
    Format data as CSV

    :param data: Data to format (dict, list of dicts, or list of lists)
    :param headers: Optional custom headers for CSV columns
    :return: CSV formatted string
    """
    output = io.StringIO()
    writer = csv.writer(output)

    # Handle different data types
    if isinstance(data, dict):
        # Single dictionary - show as key-value pairs
        if headers is None:
            headers = ["Key", "Value"]
        writer.writerow(headers)
        for key, value in data.items():
            # Serialize complex values as JSON
            val_str = json.dumps(value) if isinstance(value, (dict, list)) else str(value)
            writer.writerow([key, val_str])

    elif isinstance(data, list):
        if not data:
            return ""

        # List of dictionaries - show as CSV table
        if isinstance(data[0], dict):
            if headers is None:
                headers = list(data[0].keys())
            writer.writerow(headers)
            for item in data:
                row = []
                for h in headers:
                    value = item.get(h, '')
                    # Serialize complex values as JSON
                    val_str = json.dumps(value) if isinstance(value, (dict, list)) else str(value)
                    row.append(val_str)
                writer.writerow(row)

        # List of lists - show as CSV
        else:
            if headers:
                writer.writerow(headers)
            for item in data:
                if isinstance(item, (list, tuple)):
                    writer.writerow(item)
                else:
                    writer.writerow([item])

    else:
        # For other types, convert to string
        writer.writerow([str(data)])

    return output.getvalue()

def format_output(data: Any, output_format: str = 'json', headers: Optional[List[str]] = None) -> str:
    """
    Format data according to specified output format

    :param data: Data to format
    :param output_format: Output format (json, yaml, table, csv)
    :param headers: Optional headers for table/csv format
    :return: Formatted string
    """
    output_format = output_format.lower()

    if output_format == 'json':
        return format_json(data)
    elif output_format == 'yaml':
        return format_yaml(data)
    elif output_format == 'table':
        return format_table(data, headers)
    elif output_format == 'csv':
        return format_csv(data, headers)
    else:
        raise ValueError(f"Unsupported output format: {output_format}")

def print_formatted(data: Any, output_format: str = 'json', headers: Optional[List[str]] = None):
    """
    Print data in specified format to stdout

    :param data: Data to format and print
    :param output_format: Output format (json, yaml, table)
    :param headers: Optional headers for table format
    """
    formatted = format_output(data, output_format, headers)
    print(formatted)

File: cli/js/src/formatters.js
/**
 * Output formatters for JadeVectorDB JavaScript CLI
 *
 * Provides formatting functions for different output formats
 * (JSON, YAML, table, CSV) to improve CLI usability and integration.
 */

const yaml = require('js-yaml');
const Table = require('cli-table3');

/**
 * Format data as JSON
 */
function formatJson(data, indent = 2) {
  return JSON.stringify(data, null, indent);
}

/**
 * Format data as YAML
 */
function formatYaml(data) {
  try {
    return yaml.dump(data, {
      sortKeys: false,
      lineWidth: -1
    });
  } catch (error) {
    console.error('Warning: Error formatting YAML, falling back to JSON');
    return formatJson(data);
  }
}

/**
 * Format data as a table
 */
function formatTable(data, headers = null) {
  try {
    // Handle different data types
    if (Array.isArray(data)) {
      if (data.length === 0) {
        return 'No data to display';
      }

      // Array of objects - show as table
      if (typeof data[0] === 'object' && data[0] !== null) {
        const actualHeaders = headers || Object.keys(data[0]);
        const table = new Table({
          head: actualHeaders,
          style: { head: ['cyan'] }
        });

        data.forEach(item => {
          const row = actualHeaders.map(h => {
            const value = item[h];
            if (value === null || value === undefined) return '';
            if (typeof value === 'object') return JSON.stringify(value);
            return String(value);
          });
          table.push(row);
        });

        return table.toString();
      }

      // Array of primitives
      const table = new Table({
        head: ['Value'],
        style: { head: ['cyan'] }
      });
      data.forEach(item => table.push([String(item)]));
      return table.toString();

    } else if (typeof data === 'object' && data !== null) {
      // Single object - show as key-value pairs
      const table = new Table({
        head: ['Key', 'Value'],
        style: { head: ['cyan'] }
      });

      Object.entries(data).forEach(([key, value]) => {
        const displayValue = typeof value === 'object' ?
          JSON.stringify(value) : String(value);
        table.push([key, displayValue]);
      });

      return table.toString();
    }

    // For other types, convert to string
    return String(data);

  } catch (error) {
    console.error('Warning: Error formatting table, falling back to JSON');
    return formatJson(data);
  }
}

/**
 * Format data as CSV
 */
function formatCsv(data, headers = null) {
  try {
    const lines = [];

    // Handle different data types
    if (Array.isArray(data)) {
      if (data.length === 0) {
        return '';
      }

      // Array of objects - format as CSV table
      if (typeof data[0] === 'object' && data[0] !== null) {
        const actualHeaders = headers || Object.keys(data[0]);

        // Add header row
        lines.push(actualHeaders.map(h => escapeCSV(h)).join(','));

        // Add data rows
        data.forEach(item => {
          const row = actualHeaders.map(h => {
            const value = item[h];
            if (value === null || value === undefined) return '';
            if (typeof value === 'object') return escapeCSV(JSON.stringify(value));
            return escapeCSV(String(value));
          });
          lines.push(row.join(','));
        });
      } else {
        // Array of primitives
        lines.push('Value');
        data.forEach(item => {
          lines.push(escapeCSV(String(item)));
        });
      }

    } else if (typeof data === 'object' && data !== null) {
      // Single object - format as key-value pairs
      lines.push('Key,Value');
      Object.entries(data).forEach(([key, value]) => {
        const displayValue = typeof value === 'object' ?
          JSON.stringify(value) : String(value);
        lines.push(`${escapeCSV(key)},${escapeCSV(displayValue)}`);
      });
    } else {
      // For other types, convert to string
      return String(data);
    }

    return lines.join('\n');

  } catch (error) {
    console.error('Warning: Error formatting CSV, falling back to JSON');
    return formatJson(data);
  }
}

/**
 * Escape CSV values (handle quotes and commas)
 */
function escapeCSV(value) {
  const str = String(value);
  // If value contains comma, quote, or newline, wrap in quotes and escape quotes
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Format data according to specified output format
 */
function formatOutput(data, outputFormat = 'json', headers = null) {
  const format = outputFormat.toLowerCase();

  switch (format) {
    case 'json':
      return formatJson(data);
    case 'yaml':
      return formatYaml(data);
    case 'table':
      return formatTable(data, headers);
    case 'csv':
      return formatCsv(data, headers);
    default:
      throw new Error(`Unsupported output format: ${outputFormat}`);
  }
}

/**
 * Print formatted data to stdout
 */
function printFormatted(data, outputFormat = 'json', headers = null) {
  const formatted = formatOutput(data, outputFormat, headers);
  console.log(formatted);
}

module.exports = {
  formatJson,
  formatYaml,
  formatTable,
  formatCsv,
  formatOutput,
  printFormatted
};
